# Phase 4 — Statistical Analysis
## Dataset: Olist Brazilian E-Commerce
## Tool: Python, scipy, pandas

### Analyses Completed:
- A1: Delivery delay vs review score correlation
- A2: Freight cost distribution by category  
- A3: Hypothesis test — late vs on-time review scores
- A4: Revenue seasonality patterns
- A5: Seller outlier detection

### Key Finding:
Delivery performance is the single most statistically 
proven driver of customer satisfaction on the platform.

In [ ]:

import pandas as pd
import numpy as np
from scipy import stats
import sqlalchemy
import warnings
warnings.filterwarnings('ignore')

engine = sqlalchemy.create_engine(
    "mysql+pymysql://root:yourpassword@localhost:3306/olist_analysis"
)

# Load master table
print("Loading master table...")
df = pd.read_sql("SELECT * FROM master_orders_table", engine)
print(f" Loaded {len(df):,} rows, {len(df.columns)} columns")

# Quick sanity check
print(f"\nDelivery status breakdown:")
print(df['delivery_status'].value_counts())

Loading master table...
 Loaded 118,310 rows, 43 columns

Delivery status breakdown:
delivery_status
On Time          108163
Late               7559
Not Delivered      2588
Name: count, dtype: int64


In [ ]:
# A1: Correlation — Delivery Delay vs Review Score
# Business Question: Does delivery delay 
# directly cause lower review scores?

# Filter delivered orders only with valid scores
delivered = df[
    (df['delivery_status'] != 'Not Delivered') &
    (df['review_score'].notna()) &
    (df['delivery_delay_days'].notna())
].copy()

# Correlation calculation
correlation, p_value = stats.pearsonr(
    delivered['delivery_delay_days'],
    delivered['review_score']
)

print(f"Pearson Correlation: {correlation:.4f}")
print(f"P-value: {p_value:.6f}")
print(f"Sample size: {len(delivered):,}")
print()

# Interpretation
if p_value < 0.05:
    print(" Result is statistically significant (p < 0.05)")
else:
    print(" Result is NOT statistically significant (p > 0.05)")

if correlation < 0:
    print(f"📉 Negative correlation — as delay increases, review score decreases")
else:
    print(f"📈 Positive correlation — as delay increases, review score increases")

print()

# Average review score by delay bucket
delivered['delay_bucket'] = pd.cut(
    delivered['delivery_delay_days'],
    bins=[-float('inf'), -10, 0, 5, 10, float('inf')],
    labels=['Very Early', 'On Time', 'Slight Delay', 'Moderate Delay', 'Severe Delay']
)

bucket_summary = delivered.groupby('delay_bucket')['review_score'].agg(
    count='count',
    avg_score='mean'
).round(2)

print("Review Score by Delivery Timing:")
print(bucket_summary)


## A1 Business Insight: Delivery Delays Drive Customer Dissatisfaction

**Key Finding:**  
Delivery delay has a statistically significant negative correlation with customer review scores (**r = -0.23, p < 5001**), indicating that customer satisfaction declines as delivery delays increase.

### Impact of Delivery Delays on Reviews

- Orders delivered **on time or with minimal delay** receive an average review score of **4.16**.
- Even a **1–5 day delay** reduces the average review score to **2.94**, representing a **30% decline in customer satisfaction**.
- **Moderate and severe delays** further reduce review scores to **below 2.0**, indicating critical levels of customer dissatisfaction.

### Business Recommendation

Improving delivery performance should be a top operational priority. Since delivery delays have a direct and measurable impact on customer ratings, investments in logistics optimization, carrier performance monitoring, and proactive delivery management are likely to generate the highest return in terms of customer satisfaction and platform reputation.

In [9]:
# A2: Freight Cost Distribution by Category
# Business Question: How is freight cost 
# distributed and which categories are most 
# impacted?


# Filter delivered orders
delivered_df = df[df['order_status'] == 'delivered'].copy()

# Freight as % of price per order
delivered_df['freight_pct'] = (
    delivered_df['freight_value'] / 
    delivered_df['price'] * 100
).replace([float('inf'), -float('inf')], None).dropna()

# Overall freight distribution
print("Overall Freight Cost Distribution:")
print(f"  Mean freight %:   {delivered_df['freight_pct'].mean():.2f}%")
print(f"  Median freight %: {delivered_df['freight_pct'].median():.2f}%")
print(f"  Std deviation:    {delivered_df['freight_pct'].std():.2f}%")
print(f"  Min:              {delivered_df['freight_pct'].min():.2f}%")
print(f"  Max:              {delivered_df['freight_pct'].max():.2f}%")
print()

# By category — top 10 highest freight burden
category_freight = delivered_df.groupby('category_english').agg(
    avg_freight_pct=('freight_pct', 'mean'),
    avg_freight_value=('freight_value', 'mean'),
    avg_price=('price', 'mean'),
    total_orders=('order_id', 'count')
).round(2).sort_values('avg_freight_pct', ascending=False)

print("Top 10 Categories by Freight Burden:")
print(category_freight.head(10))
print()
print("Bottom 10 Categories by Freight Burden:")
print(category_freight.tail(10))

Overall Freight Cost Distribution:
  Mean freight %:   32.21%
  Median freight %: 23.25%
  Std deviation:    34.82%
  Min:              0.00%
  Max:              2623.53%

Top 10 Categories by Freight Burden:
                         avg_freight_pct  avg_freight_value  avg_price  \
category_english                                                         
home_comfort_2                     93.11              13.59      24.94   
dvds_blu_ray                       79.61              20.06      74.36   
christmas_supplies                 68.31              21.15      57.66   
electronics                        67.73              16.76      57.59   
flowers                            56.50              14.81      33.64   
signaling_and_security             54.82              32.69     107.49   
fashion_underwear_beach            52.03              14.48      72.10   
telephony                          50.69              15.71      70.69   
food_drink                         48.81           

## A2 Business Insight: Freight Cost Efficiency Varies Significantly Across Product Categories

**Key Finding:**  
Freight cost as a percentage of product price varies substantially across product categories, with an average freight burden of **32.21%** and a high standard deviation of **34.82%**, indicating significant differences in logistics efficiency.

### Category-Level Insights

- **Computers** have the **lowest freight burden (5.56%)**, as their high selling prices absorb shipping costs efficiently.
- Categories such as **Home Comfort 2 (93%)** and **DVDs & Blu-ray (79%)** have exceptionally high freight burdens, making them potentially **unprofitable after logistics expenses**.
- High-revenue categories including **Watches & Gifts (17.35%)** and **Health & Beauty** maintain relatively low freight costs compared to product value, making them among the **most logistics-efficient categories**.

### Business Recommendation

Optimize the product portfolio by prioritizing categories with **high revenue and low freight burden** while reviewing pricing, shipping strategies, or supplier contracts for categories where freight costs consume a large share of product value. Reducing logistics costs in high-burden categories can significantly improve overall profitability.

In [11]:
# A3: Hypothesis Testing
# Business Question: Is the review score 
# difference between late and on-time deliveries
# statistically significant?
# H0: No difference in review scores
# H1: Late deliveries have lower review scores

# Separate groups
on_time = df[
    (df['delivery_status'] == 'On Time') &
    (df['review_score'].notna())
]['review_score']

late = df[
    (df['delivery_status'] == 'Late') &
    (df['review_score'].notna())
]['review_score']

not_delivered = df[
    (df['delivery_status'] == 'Not Delivered') &
    (df['review_score'].notna())
]['review_score']

# Print group summaries
print("Group Summaries:")
print(f"  On Time     — n={len(on_time):,}  mean={on_time.mean():.3f}  std={on_time.std():.3f}")
print(f"  Late        — n={len(late):,}   mean={late.mean():.3f}  std={late.std():.3f}")
print(f"  Not Delivered — n={len(not_delivered):,}   mean={not_delivered.mean():.3f}  std={not_delivered.std():.3f}")
print()

# Mann-Whitney U Test — On Time vs Late
stat, p_value = stats.mannwhitneyu(on_time, late, alternative='greater')
print("Mann-Whitney U Test — On Time vs Late:")
print(f"  U-statistic: {stat:,.0f}")
print(f"  P-value:     {p_value:.10f}")
if p_value < 0.05:
    print("   Statistically significant — late deliveries genuinely lower scores")
else:
    print("   Not statistically significant")
print()

# Mann-Whitney U Test — Late vs Not Delivered
stat2, p_value2 = stats.mannwhitneyu(late, not_delivered, alternative='greater')
print("Mann-Whitney U Test — Late vs Not Delivered:")
print(f"  U-statistic: {stat2:,.0f}")
print(f"  P-value:     {p_value2:.10f}")
if p_value2 < 0.05:
    print("  Statistically significant — undelivered orders genuinely lower scores further")
else:
    print("   Not statistically significant")

Group Summaries:
  On Time     — n=107,493  mean=4.206  std=1.236
  Late        — n=7,368   mean=2.253  std=1.569
  Not Delivered — n=2,471   mean=1.755  std=1.324

Mann-Whitney U Test — On Time vs Late:
  U-statistic: 639,634,166
  P-value:     0.0000000000
   Statistically significant — late deliveries genuinely lower scores

Mann-Whitney U Test — Late vs Not Delivered:
  U-statistic: 10,670,024
  P-value:     0.0000000000
  Statistically significant — undelivered orders genuinely lower scores further


## A3 Business Insight: Late Deliveries Significantly Reduce Customer Satisfaction

**Key Finding:**  
The **Mann–Whitney U test** confirms a statistically significant difference in customer review scores between delivery outcomes (**p < 0.0001**). Customers receiving late deliveries consistently provide much lower ratings than those receiving orders on time.

### Statistical Evidence

- **On-time deliveries** achieve an average review score of **4.206**.
- **Late deliveries** receive an average review score of only **2.253**, representing a substantial decline in customer satisfaction.
- **Undelivered orders** perform even worse, with an average review score of **1.755**, indicating severe customer dissatisfaction.
- The extremely low **p-value (< 0.0001)** confirms that these differences are statistically significant and highly unlikely to be due to random chance.

### Business Recommendation

Delivery reliability is a critical driver of customer satisfaction. Reducing late deliveries and preventing failed deliveries should be a strategic priority through improved logistics planning, carrier performance monitoring, and proactive customer communication. These improvements can significantly increase review scores, customer trust, and long-term customer retention.


In [13]:
# A4: Revenue Seasonality Analysis
# Business Question: Are there predictable 
# revenue patterns by month and day?


delivered_df = df[df['order_status'] == 'delivered'].copy()

# Parse datetime
delivered_df['order_purchase_timestamp'] = pd.to_datetime(
    delivered_df['order_purchase_timestamp']
)

# Extract time components
delivered_df['month']     = delivered_df['order_purchase_timestamp'].dt.month
delivered_df['month_name']= delivered_df['order_purchase_timestamp'].dt.strftime('%b')
delivered_df['dayofweek'] = delivered_df['order_purchase_timestamp'].dt.day_name()
delivered_df['hour']      = delivered_df['order_purchase_timestamp'].dt.hour

# Monthly pattern
monthly = delivered_df.groupby(['month', 'month_name']).agg(
    total_orders=('order_id', 'nunique'),
    total_revenue=('price', 'sum')
).round(2).reset_index().sort_values('month')

print("Revenue by Month:")
print(monthly[['month_name', 'total_orders', 'total_revenue']].to_string(index=False))
print()

# Day of week pattern
dow_order = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']
daily = delivered_df.groupby('dayofweek').agg(
    total_orders=('order_id', 'nunique'),
    avg_order_value=('price', 'mean')
).round(2).reindex(dow_order)

print("Orders by Day of Week:")
print(daily.to_string())
print()

# Hour of day pattern
hourly = delivered_df.groupby('hour').agg(
    total_orders=('order_id', 'nunique')
).round(2)

print("Peak Order Hours:")
print(hourly.nlargest(5, 'total_orders').to_string())

Revenue by Month:
month_name  total_orders  total_revenue
       Jan          7819     1088786.12
       Feb          8208     1117338.16
       Mar          9549     1381284.11
       Apr          9101     1370222.59
       May         10295     1545226.31
       Jun          9234     1346675.85
       Jul         10031     1418177.51
       Aug         10544     1452315.33
       Sep          4151      647188.44
       Oct          4743      722375.74
       Nov          7289     1031580.23
       Dec          5514      755301.34

Orders by Day of Week:
           total_orders  avg_order_value
dayofweek                               
Monday            15701           120.41
Tuesday           15503           118.48
Wednesday         15076           119.74
Thursday          14323           119.29
Friday            13685           121.19
Saturday          10555           122.94
Sunday            11635           117.88

Peak Order Hours:
      total_orders
hour              
16          

## A4 Business Insight: Customer Purchasing Patterns Reveal Clear Temporal Trends

**Key Finding:**  
Order volume follows distinct monthly, weekly, and hourly patterns, providing valuable insights for demand forecasting, inventory planning, and marketing campaign scheduling.

### Time-Based Purchasing Trends

- Monthly order volume peaks during **May, July, and August**, with **more than 10,000 orders** each month.
- Order volume drops sharply in **September and October** to approximately **4,500 orders**. This decline is likely due to **incomplete data collection** rather than actual seasonal demand.
- **November** exhibits a noticeable increase in orders, reflecting the impact of **Black Friday promotions**.
- **Weekdays consistently outperform weekends**, with **Monday recording the highest order volume (15,701 orders)** and **Saturday the lowest (10,555 orders)**.
- Customer activity is highest between **11:00 AM and 4:00 PM**, indicating that most purchases occur during lunchtime and afternoon browsing hours.

### Business Recommendation

Align marketing campaigns, promotional offers, and inventory planning with peak demand periods. Schedule major promotions during **high-traffic weekdays (especially Monday)** and optimize advertising between **11 AM and 4 PM** to maximize customer engagement and conversion rates. Additionally, account for incomplete monthly data when performing seasonal demand analysis to avoid misleading business conclusions.

In [15]:
# A5: Seller Outlier Detection
# Business Question: Which sellers are 
# statistical outliers in performance?


delivered_df = df[df['order_status'] == 'delivered'].copy()

# Aggregate seller metrics
seller_stats = delivered_df.groupby('seller_id').agg(
    total_orders=('order_id', 'nunique'),
    total_revenue=('price', 'sum'),
    avg_review_score=('review_score', 'mean'),
    avg_freight=('freight_value', 'mean')
).round(2).reset_index()

# Filter sellers with 10+ orders
seller_stats = seller_stats[seller_stats['total_orders'] >= 10]

print(f"Total sellers with 10+ orders: {len(seller_stats):,}")
print()

# Z-scores for revenue and review score
seller_stats['revenue_zscore'] = stats.zscore(seller_stats['total_revenue'])
seller_stats['review_zscore']  = stats.zscore(seller_stats['avg_review_score'])

# Identify outliers — z-score beyond 2 std deviations
revenue_outliers = seller_stats[
    seller_stats['revenue_zscore'] > 2
].sort_values('total_revenue', ascending=False)

review_outliers = seller_stats[
    seller_stats['review_zscore'] < -2
].sort_values('avg_review_score', ascending=True)

print(f"High Revenue Outliers (z > 2): {len(revenue_outliers)}")
print(revenue_outliers[['seller_id', 'total_orders', 
                         'total_revenue', 'avg_review_score']].head(10).to_string(index=False))
print()
print(f"Low Review Score Outliers (z < -2): {len(review_outliers)}")
print(review_outliers[['seller_id', 'total_orders', 
                        'total_revenue', 'avg_review_score']].head(10).to_string(index=False))

Total sellers with 10+ orders: 1,238

High Revenue Outliers (z > 2): 34
                       seller_id  total_orders  total_revenue  avg_review_score
53243585a1d6dc2643021fd1853d8905           348      239791.94              4.12
4869f7a5dfa277a7dca6462dcf3b52b2          1124      235382.53              4.12
4a3ca9315b744ce9f8e9374361493884          1772      212145.07              3.82
fa1c13f2614d7b5c4749cbc52fecda94           578      200859.33              4.38
7c67e1448b00f6e969d365cea6b010ab           973      198334.27              3.40
7e93a43ef30c4f03f38b393420bc753a           319      172427.79              4.36
da8622b14eb17ae2831f4ac5b9dab84a          1311      170885.17              4.08
7a67c85e85bb2ce8582c35f2203ad736          1145      148662.95              4.26
1025f0e2d44d7041d6cf58b6550e0bfa           910      142882.55              3.88
955fee9216a65b617aa5c0531780ce60          1261      134070.01              4.09

Low Review Score Outliers (z < -2): 51
        

## A5 Business Insight: Seller Performance Outliers Highlight Revenue Opportunities and Quality Risks

**Key Finding:**  
Outlier analysis of **1,238 active sellers** reveals a small group of exceptionally high-performing sellers alongside a subset of consistently low-rated sellers, highlighting opportunities for growth and areas requiring immediate attention.

### Seller Performance Insights

- **34 sellers** are identified as **revenue outliers**, generating revenues more than **two standard deviations above the average**.
- The **highest-performing seller** generated approximately **R$239,791**, significantly outperforming the rest of the marketplace.
- Despite their high sales volume, these top sellers maintain **average review scores above 3.8**, demonstrating that operational scale and customer satisfaction can be achieved simultaneously.
- In contrast, **51 sellers** are identified as **customer satisfaction outliers**, with review scores **more than two standard deviations below the average**, all scoring **below 2.7**.
- Although this group represents only a small fraction of sellers, it poses a **concentrated reputational risk** to the platform due to consistently poor customer experiences.

### Business Recommendation

Use the practices of high-performing sellers as operational benchmarks to improve marketplace performance. Simultaneously, prioritize the 51 low-rated sellers for quality audits, targeted training, performance monitoring, or corrective actions to protect customer satisfaction and strengthen overall platform reputation..
